pkg/operator/controller/canary/route.go

```diff
@@ -0,0 +1,140 @@
+package canary
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/openshift/cluster-ingress-operator/pkg/manifests"
+	"github.com/openshift/cluster-ingress-operator/pkg/operator/controller"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+
+	routev1 "github.com/openshift/api/route/v1"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+)
+
+// ensureCanaryRoute ensures the canary route exists
+func (r *reconciler) ensureCanaryRoute(service *corev1.Service) (bool, *routev1.Route, error) {
+	desired := desiredCanaryRoute(service)
+	haveRoute, current, err := r.currentCanaryRoute()
+	if err != nil {
+		return false, nil, err
+	}
+
+	switch {
+	case !haveRoute:
+		if err := r.createCanaryRoute(desired); err != nil {
+			return false, nil, err
+		}
+		return r.currentCanaryRoute()
+	case haveRoute:
+		if updated, err := r.updateCanaryRoute(current, desired); err != nil {
+			return true, current, err
+		} else if updated {
+			return r.currentCanaryRoute()
+		}
+	}
+
+	return true, current, nil
+}
```

Comment:
I always find this control flow difficult to understand. Can this be simplified as follows:

```
func (r *reconciler) ensureCanaryRoute(service *corev1.Service) (bool, *routev1.Route, error) {
	desired := desiredCanaryRoute(service)

	current, err := r.currentCanaryRoute()
	if err != nil {
		if errors.IsNotFound(err) {
			return r.createCanaryRoute()
		}
		return err
	}

        // updateRouteIfNeeded performs the comparison and updates the object in
        // the API server if they differ.
	if err := r.updateRouteIfNeeded(current, desired); err != nil {
		return err
	}

	return nil
}
```

A BZ or Jira card can be created as tech debt to address the control flow in other areas of ingress/dns operators.

Reply:
This example flow leaves out the call to `r.currentCanaryRoute()` _after_ the call to `r.updateCanaryRoute`. From my understanding, it is important that we return the newest version of the resource after the update has been performed. 

Additionally, the current switch statement format is useful for situations where there is both a `haveFoo` and `wantFoo` boolean (this is the case with the ingress operator's [ensureLoadBalancer func](https://github.com/openshift/cluster-ingress-operator/blob/master/pkg/operator/controller/ingress/load_balancer_service.go#L161-L197)).

Reply:
I guess most of this is subjective. I definitely prefer the boolean based control flow, and I believe we have been trying to work towards standardizing on that format across both of our operators, but I could be wrong. @Miciah might have more context to that end. @danehans if you feel strongly about changing all of our `ensureFoo` functions to follow a new control flow, would you mind opening the BZ/jira ticket and bringing it up with the team? I'm not entirely sure if I follow what it is you have in mind.

Reply:
I find using a `switch` more readable because there are 2 Boolean variables (have X and want X) → 4 cases → 4 actions (no-op, create, delete, or update).  We have been trying to standardize this convention with the notion that maybe someday we can factor out the "ensure" logic into a dependency tree of resources, where each resource defines its "desired' function and CRUD operators.  We haven't done that refactoring, but IMO the consistency is a win.